tests/actions.rs

```diff
@@ -0,0 +1,235 @@
+    let now = Instant::now();
+    c.perform_actions(vec![null_actions].into()).await?;
+    assert!(now.elapsed().as_seconds_f64() >= 1.0);
+    Ok(())
+}
+
+async fn actions_key(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let sample_url = sample_page_url(port);
+    c.goto(&sample_url).await?;
+
+    // Test pause.
+    let key_pause = KeyActions::new("key".to_string()).pause(Duration::from_secs(1));
+    let now = Instant::now();
+    c.perform_actions(vec![key_pause].into()).await?;
+    assert!(now.elapsed().as_seconds_f64() >= 1.0);
+
+    // Test key down/up.
+    let mut elem = c.find(Locator::Id("text-input")).await?;
+    assert_eq!(elem.prop("value").await?.unwrap(), "");
+
+    let key_actions = KeyActions::new("key".to_string())
+        .then(KeyAction::Down { value: 'a' })
+        .then(KeyAction::Up { value: 'a' });
+    elem.click().await?;
+    c.perform_actions(vec![key_actions].into()).await?;
+    let mut elem = c.find(Locator::Id("text-input")).await?;
+    assert_eq!(elem.prop("value").await?.unwrap(), "a");
+    Ok(())
+}
+
+async fn actions_mouse(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let sample_url = sample_page_url(port);
+    c.goto(&sample_url).await?;
+
+    // Test pause.
+    let mouse_pause = MouseActions::new("mouse".to_string()).pause(Duration::from_secs(1));
+    let now = Instant::now();
+    c.perform_actions(vec![mouse_pause].into()).await?;
+    assert!(now.elapsed().as_seconds_f64() >= 1.0);
+
+    let elem = c.find(Locator::Id("button-alert")).await?;
+
+    // Test mouse down/up.
+    let mouse_actions = MouseActions::new("mouse".to_string())
+        .then(PointerAction::MoveToElement {
+            element: elem,
+            duration: None,
+            x: 0,
+            y: 0,
+        })
+        .then(PointerAction::Down {
+            button: MOUSE_BUTTON_LEFT,
+        })
+        .then(PointerAction::Up {
+            button: MOUSE_BUTTON_LEFT,
+        });
+
+    c.perform_actions(vec![mouse_actions].into()).await?;
+    assert_eq!(c.get_alert_text().await?, "This is an alert");
+    c.dismiss_alert().await?;
+    Ok(())
+}
+
+async fn actions_mouse_move(mut c: Client, port: u16) -> Result<(), error::CmdError> {
+    let sample_url = sample_page_url(port);
+    c.goto(&sample_url).await?;
+
+    let mut elem = c.find(Locator::Id("text-input")).await?;
+    let rect = elem.rectangle().await?;
+    let elem_center_y = rect.1 + (rect.3 / 2.0);
+
+    elem.send_keys("fantoccini").await?;
+    assert_eq!(elem.prop("value").await?.unwrap(), "fantoccini");
+
+    // Test mouse MoveTo and MoveBy, by implementing drag-and-drop to select text
+    // in the text input element.
+    let mouse_actions = MouseActions::new("mouse".to_string())
+        // Move to the left edge of the input element.
+        // Offset by 1 pixel to ensure we click inside the element.
+        .then(PointerAction::MoveTo {
```

Comment:
Won't this break if the text ends up wrapping? Do we know the text element will be sufficiently wide?

Reply:
Since this is running in a fairly specific test environment I feel like it should be fine.
Just to be more sure I've added specific styling to the text box to enforce a specific width and font size.